numba/cuda/tests/cudapy/test_lineinfo.py

```diff
@@ -1,42 +1,207 @@
+        self._check(foo, sig=(int32[:],), expect=False)
 
     def test_lineinfo_in_asm(self):
+        if not NVVM().is_nvvm70:
+            self.skipTest("lineinfo not generated for NVVM 3.4")
+
         @cuda.jit(lineinfo=True)
         def foo(x):
             x[0] = 1
 
-        self._check(foo, sig=(types.int32[:],), expect=True)
+        self._check(foo, sig=(int32[:],), expect=True)
+
+    def test_lineinfo_maintains_error_model(self):
+        sig = (float32[::1], float32[::1])
+
+        @cuda.jit(sig, lineinfo=True)
+        def divide_kernel(x, y):
+            x[0] /= y[0]
+
+        llvm = divide_kernel.inspect_llvm(sig)
+
+        # When the error model is Python, the device function returns 1 to
+        # signal an exception (e.g. divide by zero) has occurred. When the
+        # error model is the default NumPy one (as it should be when only
+        # lineinfo is enabled) the device function always returns 0.
+        self.assertNotIn('ret i32 1', llvm)
+
+    def test_no_lineinfo_in_device_function(self):
+        # Ensure that no lineinfo is generated in device functions by default.
+        @cuda.jit
+        def callee(x):
+            x[0] += 1
+
+        @cuda.jit
+        def caller(x):
+            x[0] = 1
+            callee(x)
+
+        sig = (int32[:],)
+        self._check(caller, sig=sig, expect=False)
+
+    def test_lineinfo_in_device_function(self):
+        if not NVVM().is_nvvm70:
+            self.skipTest("lineinfo not generated for NVVM 3.4")
+
+        # First we define a device function / kernel pair and run the usual
+        # checks on the generated LLVM and PTX.
+
+        @cuda.jit(lineinfo=True)
+        def callee(x):
+            x[0] += 1
+
+        @cuda.jit(lineinfo=True)
+        def caller(x):
+            x[0] = 1
+            callee(x)
+
+        sig = (int32[:],)
+        self._check(caller, sig=sig, expect=True)
+
+        # Now we can check the PTX of the device function specifically.
+
+        ptx = caller.inspect_asm(sig)
+        ptxlines = ptx.splitlines()
+
+        # To check the device function, we need to identify its boundaries.
+
+        # A line beginning with ".weak .func"
+        devfn_start = re.compile(r'^\.weak\s*\.func')
+
+        # Identify the beginning of the function.
+        start = None
+
+        for lineno, line in enumerate(ptxlines):
+            if devfn_start.match(line) is not None:
+                # We will begin our search on the line following the
+                # declaration
+                start = lineno + 1
+
+        if start is None:
+            self.fail(f'Could not identify device function in:\n\n{ptx}')
+
+        # Identify the end of the function
+        end = None
+
+        for offset, line in enumerate(ptxlines[start:]):
+            # Assume the end of the function is a line with an unindented '}'
+            if line[:1] == '}':
+                end = start + offset
```

Comment:
Should there be a `break` here once the match is found?

Reply:
I think we might as well do that here for the same reason as above.

Reply:
This code is now removed as it is no longer needed anyway. (As per https://github.com/numba/numba/pull/8594#discussion_r1108670106)